include/openssl/type_check.h

```diff
@@ -63,17 +63,24 @@
 extern "C" {
 #endif
 
-
-#if defined(__cplusplus) || (defined(_MSC_VER) && !defined(__clang__))
-// In C++ and non-clang MSVC, |static_assert| is a keyword.
-#define OPENSSL_STATIC_ASSERT(cond, msg) static_assert(cond, msg)
-#else
-// C11 defines the |_Static_assert| keyword and the |static_assert| macro in
-// assert.h. While the former is available at all versions in Clang and GCC, the
-// later depends on libc and, in glibc, depends on being built in C11 mode. We
-// do not require this, for now, so use |_Static_assert| directly.
-#define OPENSSL_STATIC_ASSERT(cond, msg) _Static_assert(cond, msg)
-#endif
+// Previously we defined |OPENSSL_STATIC_ASSERT| to use one of two keywords:
+// |Static_assert| or |static_assert|. The latter was used if we were compiling
+// a C++ translation unit or on Windows (excluding when using a Clang compiler).
+// The former was used in other cases. However, these two keywords are not
+// defined before C11. So, we can't rely on these when we want to be C99
+// compliant. If we at some point decides that we want to only be compliant with
+// C11 (and up), we can reintroduce these keywords. Instead, use a method that
+// is guaranteed to be C99 compliant and still gives use an equivalent static
+// assert mechanism.
+#define AWSLC_CONCAT(left, right) left##right
+#define AWSLC_STATIC_ASSERT_DEFINE(cond, msg) typedef struct { \
+        unsigned int AWSLC_CONCAT(static_assertion_, msg) : (cond) ? 1 : - 1; \
+    } AWSLC_CONCAT(static_assertion_, msg) OPENSSL_UNUSED
```

Comment:
Could you please add an explanation of how this works?

And another question, did you consider using the aws-c-common solution, is there any trade-off between the two solutions?

Reply:
The width element of a bit field cannot be negative.

I guess I was easily persuaded by this line: https://github.com/Perl/perl5/blob/5640a370e8b19af74b8ca0b4694464c21a87916b/perl.h#L3860
But I don't really see why that matters. I also thought I could get around unused errors. But I couldn't, and in the end, I just stuck with the solution I had.

Reply:
> The width element of a bit field cannot be negative.

Thanks, but I thought maybe to add this comment in the code :)

> I guess I was easily persuaded by this line: https://github.com/Perl/perl5/blob/5640a370e8b19af74b8ca0b4694464c21a87916b/perl.h#L3860
But I don't really see why that matters. I also thought I could get around unused errors. But I couldn't, and in the end, I just stuck with the solution I had.

Thanks for the link, that explains it.